File: src/lib/localization.ts
import { DayPrices, FuelType, Price } from './prices'

export type Language = 'da' | 'en'

export function getErrorText(language: Language): string {
  switch (language) {
    case 'da':
      return 'Der blev ikke fundet nogen priser for den dato'
    case 'en':
      return 'No prices were found for that date'
    default:
      return 'No prices were found for that date'
  }
}

export function getText(
  price: DayPrices,
  fuelType: FuelType,
  language: Language,
): string {
  switch (language) {
    case 'da':
      return getTextDanish(price, fuelType)
    case 'en':
      return getTextEnglish(price, fuelType)
    default:
      return getTextEnglish(price, fuelType)
  }
}

function getTextDanish(price: DayPrices, fuelType: FuelType): string {
  const { kroner, orer } = priceToKronerAndOrer(price.today)
  let text = `${fuelTypeToString(
    fuelType,
    'da',
  )} koster ${kroner} kroner og ${orer} ører i dag.`

  if (price.yesterday) {
    const { kroner, orer } = priceToKronerAndOrer(price.yesterday)
    let diffTxt = 'den samme'
    if (price.yesterday.price > price.today.price) {
      diffTxt = 'højere'
    } else if (price.yesterday.price < price.today.price) {
      diffTxt = 'lavere'
    }
    text = `${text} I går var prisen ${diffTxt}: ${kroner} kroner og ${orer} ører.`
  }
  if (price.tomorrow) {
    const { kroner, orer } = priceToKronerAndOrer(price.tomorrow)
    let diffTxt = 'den samme'
    if (price.tomorrow.price > price.today.price) {
      diffTxt = 'højere'
    } else if (price.tomorrow.price < price.today.price) {
      diffTxt = 'lavere'
    }
    text = `${text} I morgen vil prisen være ${diffTxt}: ${kroner} kroner og ${orer} ører.`
  }
  return text
}

function getTextEnglish(price: DayPrices, fuelType: FuelType): string {
  let text = `Today, the price of ${fuelTypeToString(fuelType, 'en')} is ${
    price.today.price
  } kroner.`
  if (price.yesterday) {
    let diffTxt = 'the same'
    if (price.yesterday.price > price.today.price) {
      diffTxt = 'higher'
    } else if (price.yesterday.price < price.today.price) {
      diffTxt = 'lower'
    }
    text = `${text} Yesterday the price was ${diffTxt}: ${price.yesterday.price} kroner.`
  }
  if (price.tomorrow) {
    let diffTxt = 'the same'
    if (price.tomorrow.price > price.today.price) {
      diffTxt = 'higher'
    } else if (price.tomorrow.price < price.today.price) {
      diffTxt = 'lower'
    }
    text = `${text} Tomorrow the price will be ${diffTxt}: ${price.tomorrow.price} kroner.`
  }
  return text
}

function priceToKronerAndOrer(price: Price): { kroner: string; orer: string } {
  const parts = price.price.toString().split('.')
  return {
    kroner: parts[0],
    orer: parts[1],
  }
}

function fuelTypeToString(fuelType: FuelType, language: Language): string {
  switch (language) {
    case 'da': {
      switch (fuelType) {
        case 'Diesel':
          return 'Diesel'
        case 'Octane100':
          return 'Oktan 100'
        case 'Unleaded95':
          return 'Blyfri oktan 95'
        default:
          return 'Blyfri oktan 95'
      }
    }
    case 'en': {
      switch (fuelType) {
        case 'Diesel':
          return 'Diesel'
        case 'Octane100':
          return 'Octane 100'
        case 'Unleaded95':
          return 'Unleaded octane 95'
        default:
          return 'Unleaded octane 95'
      }
    }
  }
}

File: services/fuelpricesapi/localization.go
package main

import (
	"fmt"
	"log"
	"strconv"
	"strings"
)

type Language string

const (
	LangDa Language = "da"
	LangEn Language = "en"
)

func (l Language) GetErrorText() string {
	switch l {
	case LangDa:
		return "Der blev ikke fundet priser for den dato"
	default:
		return "No prices were found for that date"
	}
}

func (l Language) GetText(prices *DayPrices, fuelType FuelType) string {
	switch l {
	case LangDa:
		return getTextDanish(prices, fuelType)
	default:
		return getTextEnglish(prices, fuelType)
	}
}

func getTextEnglish(prices *DayPrices, fuelType FuelType) string {
	lang := LangEn
	text := fmt.Sprintf("Today, the price of %v is %.2f kroner.", lang.fuelTypeString(fuelType), prices.Today.Price)
	if prices.Yesterday != nil && prices.Yesterday.Price > 0 {
		diffText := lang.getDiffText(prices.Today, prices.Yesterday)
		text = fmt.Sprintf("%v Yesterday the price was %v: %.2f kroner.", text, diffText, prices.Yesterday.Price)
	}
	if prices.Tomorrow != nil && prices.Tomorrow.Price > 0 {
		diffText := lang.getDiffText(prices.Today, prices.Tomorrow)
		text = fmt.Sprintf("%v Tomorrow the price will be %v: %.2f kroner.", text, diffText, prices.Tomorrow.Price)
	}
	return text
}

func getTextDanish(prices *DayPrices, fuelType FuelType) string {
	lang := LangDa
	kroner, orer, err := priceToKronerAndOrer(prices.Today)
	if err != nil {
		log.Printf("failed to convert today price to kroner and orer: %v", err)
		return lang.GetErrorText()
	}

	text := fmt.Sprintf("%v koster %v kroner og %v ører i dag.", lang.fuelTypeString(fuelType), kroner, orer)
	if prices.Yesterday.Price > 0 {
		kroner, orer, err = priceToKronerAndOrer(prices.Yesterday)
		if err != nil {
			log.Printf("failed to convert yesterday price to kroner and orer: %v", err)
			return lang.GetErrorText()
		}
		diffText := lang.getDiffText(prices.Today, prices.Yesterday)
		text = fmt.Sprintf("%v I går var prisen %v: %v kroner og %v ører.", text, diffText, kroner, orer)
	}
	if prices.Tomorrow.Price > 0 {
		kroner, orer, err = priceToKronerAndOrer(prices.Tomorrow)
		if err != nil {
			log.Printf("failed to convert tomorrow price to kroner and orer: %v", err)
			return lang.GetErrorText()
		}
		diffText := lang.getDiffText(prices.Today, prices.Tomorrow)
		text = fmt.Sprintf("%v I morgen vil prisen være %v: %v kroner og %v ører.", text, diffText, kroner, orer)
	}
	return text
}

func (l Language) getDiffText(today *Price, otherDay *Price) string {
	switch l {
	case LangDa:
		if otherDay.Price > today.Price {
			return "højere"
		} else if otherDay.Price < today.Price {
			return "lavere"
		} else {
			return "den samme"
		}
	default:
		if otherDay.Price > today.Price {
			return "higher"
		} else if otherDay.Price < today.Price {
			return "lower"
		} else {
			return "the same"
		}
	}
}

func priceToKronerAndOrer(price *Price) (kroner string, orer string, err error) {
	parts := strings.Split(fmt.Sprintf("%f", price.Price), ".")
	if len(parts) != 2 {
		return "0", "0", fmt.Errorf("failed to parse price")
	}
	kronerInt, err := strconv.Atoi(parts[0])
	if err != nil {
		return "0", "0", fmt.Errorf("failed to parse kroner: %w", err)
	}
	kroner = strconv.Itoa(kronerInt)
	orer = strings.TrimPrefix(parts[1][0:2], "0")
	return kroner, orer, nil
}

func (l Language) fuelTypeString(fuelType FuelType) string {
	switch l {
	case LangDa:
		switch fuelType {
		case FuelTypeOctane100:
			return "Oktan 100"
		case FuelTypeDiesel:
			return "Diesel"
		default:
			return "Blyfri oktan 95"
		}
	default:
		switch fuelType {
		case FuelTypeOctane100:
			return "Oktan 100"
		case FuelTypeDiesel:
			return "Diesel"
		default:
			return "Unleaded octane 95"
		}
	}
}
